wtxmgr/tx_test.go

```diff
@@ -2355,3 +2355,338 @@ func TestTxLabel(t *testing.T) {
+			// manually unlock it.
+			name: "unlock output",
+			run: func(t *testing.T, s *Store, ns walletdb.ReadWriteBucket) {
+				lock(t, s, ns, 1, confirmedOutPoint, nil)
+				unlock(
+					t, s, ns, 2, confirmedOutPoint,
+					ErrOutputUnlockNotAllowed,
+				)
+				unlock(t, s, ns, 1, confirmedOutPoint, nil)
+			},
+		},
+		{
+			// Asserts that locking an output that's already locked
+			// with the correct ID results in an extension of the
+			// lock.
+			name: "extend locked output lease",
+			run: func(t *testing.T, s *Store, ns walletdb.ReadWriteBucket) {
+				// Lock the output and wait for half the
+				// duration. It should remain locked.
+				lock(t, s, ns, 1, confirmedOutPoint, nil)
+				<-time.After(lockDuration / 2)
+				assertLocked(t, ns, confirmedOutPoint, true)
+
+				// Lock it once again, extending its expiration,
+				// and wait for half the duration. It should
+				// remain locked.
+				lock(t, s, ns, 1, confirmedOutPoint, nil)
+				<-time.After(lockDuration / 2)
+				assertLocked(t, ns, confirmedOutPoint, true)
+
+				// Wait for half the duration again, it should
+				// now be unlocked.
+				<-time.After(lockDuration / 2)
+				assertLocked(t, ns, confirmedOutPoint, false)
+			},
+		},
+		{
+			// Asserts that balances are reflected properly after
+			// locking confirmed and unconfirmed outputs.
+			name: "balance after locked outputs",
+			run: func(t *testing.T, s *Store, ns walletdb.ReadWriteBucket) {
+				// We should see our full balance before locking
+				// any outputs.
+				assertBalance(
+					t, s, ns, false, block.Height, balance,
+				)
+
+				// Lock all of our outputs. Our balance should
+				// be 0.
+				lock(t, s, ns, 1, unconfirmedOutPoint, nil)
+				lock(t, s, ns, 1, confirmedOutPoint, nil)
+				assertBalance(t, s, ns, false, block.Height, 0)
+
+				// Wait for the output locks to expire, causing
+				// our full balance to return .
+				<-time.After(lockDuration)
+				assertBalance(
+					t, s, ns, false, block.Height, balance,
+				)
+			},
+		},
+		{
+			// Asserts that the available utxos are reflected
+			// properly after locking confirmed and unconfirmed
+			// outputs.
+			name: "utxos after locked outputs",
+			run: func(t *testing.T, s *Store, ns walletdb.ReadWriteBucket) {
+				// We should see all of our utxos before locking
+				// any.
+				assertUtxos(t, s, ns, []wire.OutPoint{
+					unconfirmedOutPoint,
+					confirmedOutPoint,
+				})
+
+				// Lock the unconfirmed utxo, we should now only
+				// see the confirmed.
+				lock(t, s, ns, 1, unconfirmedOutPoint, nil)
+				assertUtxos(t, s, ns, []wire.OutPoint{
+					confirmedOutPoint,
+				})
+
+				// Now lock the confirmed utxo, we should no
+				// longer see any utxos available.
+				lock(t, s, ns, 1, confirmedOutPoint, nil)
+				assertUtxos(t, s, ns, nil)
+
+				// Wait for the output locks to expire for the
+				// utxos to become available once again.
+				<-time.After(lockDuration)
```

Comment:
If time is mocked out, these could just be insta sends, which'll speed up the tests.